src/matchmaker.rs

```diff
@@ -0,0 +1,358 @@
+
+        if self.matchmaker.query.len() != 0 {
+            self.matchmaker.query.push(' ')
+        }
+
+        match query_type {
+            QueryType::Term(term) => self
+                .matchmaker
+                .query
+                .push_str(format!("{}{}{}{}", boolean, field, term, boost).as_str()),
+            QueryType::Range { operator, value } => {
+                let op = match operator {
+                    RangeOperator::GT => ">",
+                    RangeOperator::LT => "<",
+                    RangeOperator::GEQ => ">=",
+                    RangeOperator::LEQ => "<=",
+                };
+                self.matchmaker
+                    .query
+                    .push_str(format!("{}{}{}{}{}", boolean, field, op, value, boost).as_str())
+            }
+        }
+
+        self.matchmaker
+    }
+}
+
+impl<'a> Matchmaker {
+    pub fn new() -> Self {
+        Matchmaker {
+            min_count: 2,
+            max_count: 100,
+            string_properties: HashMap::new(),
+            numeric_properties: HashMap::new(),
+            query: "".to_owned(),
+        }
+    }
+
+    pub fn string_properties(&self) -> String {
+        let mut str = "{".to_owned();
+
+        let mut properties = self.string_properties.iter().map(|property| {
+            format!("\"{}\": \"{}\"", property.0, property.1)
+        }).collect::<Vec<String>>();
+
+        // The tests need to have a deterministic order
+        #[cfg(test)]
+        properties.sort();
+
+        str += &properties.join(",");
+
+        str += "}";
+        str
+    }
+
+    pub fn numeric_properties(&self) -> String {
+        let mut str = "{".to_owned();
+
+        let mut properties = self.numeric_properties.iter().map(|property| {
+            format!("\"{}\": {}", property.0, property.1)
+        }).collect::<Vec<String>>();
+
+        // The tests need to have a deterministic order
+        #[cfg(test)]
+        properties.sort();
+
+        str += &properties.join(",");
+
+        str += "}";
+        str
+    }
+
+    pub fn property_exists(&self, name: &str) -> bool {
+        self.string_properties.contains_key(name) || self.numeric_properties.contains_key(name)
+    }
+
+    pub fn min(&mut self, min: u32) -> &mut Self {
+        self.min_count = min;
+        self
+    }
+
+    pub fn max(&mut self, max: u32) -> &mut Self {
+        self.max_count = max;
+        self
+    }
+
+    pub fn add_string_property(&mut self, name: &str, value: &str) -> &mut Self {
+        assert!(!self.property_exists(name));
+
+        self.string_properties
+            .insert(name.to_owned(), value.to_owned());
+
+        self
+    }
+
+    pub fn add_numeric_property(&mut self, name: &str, value: f64) -> &mut Self {
+        assert!(!self.property_exists(name));
+
+        self.numeric_properties.insert(name.to_owned(), value);
+
+        self
+    }
+
+    pub fn add_query_item(&'a mut self, property: &'a str) -> QueryItemBuilder<'a> {
```

Comment:
Here `add_numeric_property` and `add_query_item` have very similar name but very different semantics, I would consider using QueryItemBuilder as an argument instead (part of this comment: https://github.com/erlend-sh/nakama-rs/pull/5/files#r611120506)

Than compiler will give way nicer error message for situation like this 
`matchmaker.add_query_item("").add_numeric_property("", 123)`

And it will remove the confusion on how `matchmaker.something().build().something().build()` is even possible